### PR TITLE
Split the rename function according to the level and pass expression by copy instead of reference

### PR DESCRIPTION
**Pull request by romainbrenguier**
~This is based on https://github.com/diffblue/cbmc/pull/3988 and~ This is an intermediary step towards https://github.com/diffblue/cbmc/pull/3986
We split the rename functions according to there level, so that we can later make them return types which reflects the transformation.

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [ ] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [na] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [na] My commit message includes data points confirming performance improvements (if claimed).
- [ ] My PR is restricted to a single feature or bugfix.
- [x] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

File: src/goto-symex/goto_symex_state.cpp [hunk 1]
```diff
@@ -280,14 +280,12 @@ static void set_l2_indices(
   ssa_expr.set_level_2(level2.current_count(ssa_expr.get_identifier()));
 }
 
-void goto_symex_statet::rename_level0(exprt &expr, const namespacet &ns)
+exprt goto_symex_statet::rename_level0(exprt expr, const namespacet &ns)
```
Comment (tautschnig) [inline src/goto-symex/goto_symex_state.cpp]: The parameter should be `const exprt &expr`

File: src/goto-symex/goto_symex_state.cpp [hunk 2]
```diff
@@ -280,6 +280,46 @@ static void set_l2_indices(
   ssa_expr.set_level_2(level2.current_count(ssa_expr.get_identifier()));
 }
 
+void goto_symex_statet::rename_level0(exprt &expr, const namespacet &ns)
```
Comment (tautschnig) [inline src/goto-symex/goto_symex_state.cpp]: I would expect that adding this function goes together with removing the same code from another function. As-is, it's difficult to assess whether this code does the same as the old one.

File: src/goto-symex/goto_symex_state.cpp (hunk 1 above)
Comment (romainbrenguier) [reply src/goto-symex/goto_symex_state.cpp]: In the follow up PR I will make this return another type, and then having the argument passed by copy allow to move the argument, potentially saving a copy. I also refactored the code so that this will not get called recursively.

**Comment by romainbrenguier** [comment]
@tautschnig @smowton this is ready for review

File: src/goto-symex/goto_symex_state.cpp [hunk 3]
```diff
@@ -261,97 +262,160 @@ void goto_symex_statet::set_l2_indices(
...
+  }
+}
 
-    expr=ssa_exprt(expr);
-    rename(expr, ns, level);
+void goto_symex_statet::rename_level0(exprt &expr, const namespacet &ns)
+{
+  renaming_strategyt strategy;
+  strategy.rename_ssa = [this](ssa_exprt ssa, const namespacet &ns) -> exprt {
+    set_l0_indices(ssa, ns);
```
Comment (smowton) [inline src/goto-symex/goto_symex_state.cpp]: Surely `rename_level0_ssa`

**Review by smowton — COMMENTED** [review]
The strategy pattern seems very verbose compared to just passing `levelt`, since 2/3 of its members are always the same and the other one only ever differs by level.

File: src/goto-symex/goto_symex_state.cpp (hunk 3 above)
Comment (romainbrenguier) [reply src/goto-symex/goto_symex_state.cpp]: done

**Comment by romainbrenguier** [comment]
> The strategy pattern seems very verbose compared to just passing `levelt`, 

True this is more verbose, the main intention was to avoid late decision making in the rename function, we already know what the level is, so there is no reason to add more ifs in the loop. 
I also find it more structured that way but that's a matter of taste. 

> since 2/3 of its members are always the same and the other one only ever differs by level.

Note that the members of the strategy are not exactly the same since the level argument differs, and we could think of also splitting the called functions according to their level.

File: src/goto-symex/goto_symex_state.cpp (hunk 3 above)
Comment (romainbrenguier) [reply src/goto-symex/goto_symex_state.cpp]: done

**Review by tautschnig — CHANGES_REQUESTED** [review]
I have only started reviewing this and I'm afraid the first commit is a deal breaker: "Refactor goto_symex_state::rename to use a strategy" - there are some very good changes in there (such as nicely introducing the depth iterator), but these are completely unrelated to introducing a strategy. If such changes are a necessary precondition to introducing a strategy then they must go in a separate commit, quite possibly even a separate PR. This is an area of code that is crucial both in terms of performance and correctness. We must not have multiple changes in a single commit as it prohibits identifying both performance and correctness issues or advances. Yes, we also need to be able to pinpoint improvements. Let's make such changes in isolation, benchmark them, and then follow up with the next change.

**Comment by romainbrenguier** [comment]
@tautschnig 
> I have only started reviewing this and I'm afraid the first commit is a deal breaker: "Refactor goto_symex_state::rename to use a strategy" - there are some very good changes in there (such as nicely introducing the depth iterator), but these are completely unrelated to introducing a strategy. If such changes are a necessary precondition to introducing a strategy then they must go in a separate commit, quite possibly even a separate PR. This is an area of code that is crucial both in terms of performance and correctness. We must not have multiple changes in a single commit as it prohibits identifying both performance and correctness issues or advances. Yes, we also need to be able to pinpoint improvements. Let's make such changes in isolation, benchmark them, and then follow up with the next change.

That's a good idea indeed. I tried to split the first commit, and started to profile the changes. It turns out changing the recursive function into one using iterators is not such a good idea. Just making it non-recursive (without the use of strategies) makes it significantly slower. I thought I would be able to save expression writes that way, but I think `mutate` is an issue here as it will detach all the expressions in the stack for each mutate. 
I think I will redo these changes, keeping the function recursive, and using a template argument instead of strategies.
If we want to minimize the number of expression write, I think we need to make `rename(typet ...)` return an optional so that we know wether it needs to be updated.

**Comment by tautschnig** [comment]
> I tried to split the first commit, and started to profile the changes. It turns out changing the recursive function into one using iterators is not such a good idea.

Unless you have binned it already and `git reflog` can't recover it either, would you still be ok to create a pull request just with this change? I'd at least try to discuss it. Notably, the cost may also just be distributed differently.

**Comment by romainbrenguier** [comment]
@tautschnig 
> Unless you have binned it already and git reflog can't recover it either, would you still be ok to create a pull request just with this change? I'd at least try to discuss it. Notably, the cost may also just be distributed differently.

Pull request here: https://github.com/diffblue/cbmc/pull/4254

**Comment by romainbrenguier** [comment]
@tautschnig I rewrote the PR to use templates instead of strategies. This is ready for review.

**Review by allredj — COMMENTED** [review]
:no_entry_sign:
This PR failed Diffblue compatibility checks (cbmc commit: 659d220c20b55d5e17136af346b5aed6d2b63cc0).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/102171205
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

The incompatibility may have been introduced by an earlier PR. In that case merging this
PR should be avoided unless it fixes the current incompatibility.

**Review by smowton — APPROVED** [review]
This looks great -- one suggestion, since a fair majority of renamings are to L2, maybe use a default template parameter much as there is currently a default ordinary parameter?

File: src/goto-symex/symex_decl.cpp [hunk 4]
```diff
@@ -69,7 +69,11 @@ void goto_symext::symex_decl(statet &state, const symbol_exprt &expr)
   symex_renaming_levelt::increase_counter(level2_it);
   const bool record_events=state.record_events;
   state.record_events=false;
-  state.rename<goto_symex_statet::L2>(ssa, ns);
+  exprt expr_l2 = state.rename<statet::L2>(std::move(ssa), ns);
+  INVARIANT(
+    expr_l2.id() == ID_symbol && expr_l2.get_bool(ID_C_SSA_symbol),
+    "symbol to declare should not be replaced by constant propagation");
```
Comment (tautschnig) [inline src/goto-symex/symex_decl.cpp:75]: Fair enough, but unrelated to the other changes here?

File: src/goto-symex/symex_builtin_functions.cpp [hunk 5]
```diff
@@ -481,11 +475,8 @@ void goto_symext::symex_trace(
     irep_idt event = to_string_constant(code.arguments()[1].op0()).get_value();
 
     for(std::size_t j=2; j<code.arguments().size(); j++)
-    {
-      exprt var(code.arguments()[j]);
-      state.rename<goto_symex_statet::L2>(var, ns);
-      vars.push_back(var);
-    }
+      vars.push_back(
+        state.rename<goto_symex_statet::L2>(code.arguments()[j], ns));
```
Comment (tautschnig) [inline src/goto-symex/symex_builtin_functions.cpp]: Nit pick: keep the braces.

File: src/goto-symex/symex_builtin_functions.cpp [hunk 6]
```diff
@@ -359,18 +355,16 @@ void goto_symext::symex_output(
...
   std::list<exprt> args;
 
   for(std::size_t i=1; i<code.operands().size(); i++)
   {
     args.push_back(code.operands()[i]);
-    state.rename<goto_symex_statet::L2>(args.back(), ns);
-    do_simplify(args.back());
+    exprt l2_arg = state.rename<goto_symex_statet::L2>(args.back(), ns);
+    do_simplify(l2_arg);
+    args.back() = std::move(l2_arg);
```
Comment (tautschnig) [inline src/goto-symex/symex_builtin_functions.cpp]: I think you should get rid of `args.push_back(code.operands()[i])`

File: src/goto-symex/symex_builtin_functions.cpp [hunk 7]
```diff
@@ -336,17 +333,16 @@ void goto_symext::symex_input(
...
```
Comment (tautschnig) [inline src/goto-symex/symex_builtin_functions.cpp]: Get rid of `args.push_back(...)`

File: src/goto-symex/symex_builtin_functions.cpp [hunk 8]
```diff
@@ -311,11 +309,10 @@ void goto_symext::symex_printf(
 {
   PRECONDITION(!rhs.operands().empty());
 
-  exprt tmp_rhs=rhs;
-  state.rename<goto_symex_statet::L2>(tmp_rhs, ns);
+  exprt tmp_rhs = state.rename<goto_symex_statet::L2>(rhs, ns);
   do_simplify(tmp_rhs);
 
-  const exprt::operandst &operands=tmp_rhs.operands();
+  const exprt::operandst &operands = tmp_rhs.operands();
```
Comment (tautschnig) [inline src/goto-symex/symex_builtin_functions.cpp]: Unnecessary change.

File: src/goto-symex/goto_symex_state.cpp [hunk 9]
```diff
@@ -727,7 +724,10 @@ void goto_symex_statet::rename(
     {
       // be careful, or it might get cyclic
       if(component.type().id() == ID_array)
-        rename<level>(to_array_type(component.type()).size(), ns);
+      {
+        auto &array_type = to_array_type(component.type());
+        array_type.size() = rename<level>(array_type.size(), ns);
```
Comment (tautschnig) [inline src/goto-symex/goto_symex_state.cpp]: Above you use `std::move(array_type.size())` - please be consistent.

**Review by tautschnig — CHANGES_REQUESTED** [review]
I think this mostly looks good, but I strongly dislike that users now have two seemingly very different APIs to do L0/L1 renaming. Couldn't we have `template<T> ssa_exprt rename(ssa_exprt, namespacet)` so that the compiler would just choose the most appropriate variant for you?

File: src/goto-symex/symex_decl.cpp (hunk 4 above)
Comment (romainbrenguier) [reply src/goto-symex/symex_decl.cpp:75]: We need these conditions to be sure `to_ssa_expr` will not failed in the next line. Before `ssa` was modified in-place so there was no conversion here.

**Comment by romainbrenguier** [comment]
> I think this mostly looks good, but I strongly dislike that users now have two seemingly very different APIs to do L0/L1 renaming. Couldn't we have `template<T> ssa_exprt rename(ssa_exprt, namespacet)` so that the compiler would just choose the most appropriate variant for you?

I think this will make it ambiguous because `ssa_exprt` is also an `exprt`, but I guess nothing wrong can happen. I will try your solution.

File: src/goto-symex/symex_decl.cpp (hunk 4 above)
Comment (tautschnig) [reply src/goto-symex/symex_decl.cpp:75]: Ah, indeed that possible problem was hidden before. But `to_ssa_expr` should already do the same checking for you?

**Comment by romainbrenguier** [comment]
@tautschnig 
> I think this mostly looks good, but I strongly dislike that users now have two seemingly very different APIs to do L0/L1 renaming. Couldn't we have template<T> ssa_exprt rename(ssa_exprt, namespacet) so that the compiler would just choose the most appropriate variant for you?

> I think this will make it ambiguous because ssa_exprt is also an exprt, but I guess nothing wrong can happen. I will try your solution.

The problem is that there is no equivalent of this function for L2 (it would not always return an ssa_exprt  because of constant propagation). So either we will get an template that only works for level=L0, or L1 which I think is confusing. Or we need to return a type which depends on the level, which will involve a bit more template coding. What do you think?

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 3333db08bbfcfdb1603fb87dd8f328c4fe3ed457).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/102350102

**Comment by tautschnig** [comment]
> The problem is that there is no equivalent of this function for L2 (it would not always return an ssa_exprt because of constant propagation). So either we will get an template that only works for level=L0, or L1 which I think is confusing. Or we need to return a type which depends on the level, which will involve a bit more template coding. What do you think?

Isn't it possible to delete a specialization? I think you should be able to do
````c++
template<> ssa_exprt rename<L2>(ssa_exprt, const namespacet &) = delete;
````
and the compiler should then happily use `exprt rename<L2>(...)` even when you pass in an `ssa_exprt` as first argument. But maybe I'm too optimistic about C++ here?

**Comment by romainbrenguier** [comment]
@tautschnig 
> Isn't it possible to delete a specialization? I think you should be able to do
> 
 ```c++
template<> ssa_exprt rename<L2>(ssa_exprt, const namespacet &) = delete;
```
> and the compiler should then happily use `exprt rename<L2>(...)` even when you pass in an `ssa_exprt` as first argument. But maybe I'm too optimistic about C++ here? use of deleted function 

In that case I get the error: `use of deleted function ‘ssa_exprt goto_symex_statet::rename(ssa_exprt, const namespacet&) [with goto_symex_statet::levelt level = (goto_symex_statet::levelt)2u]’`
when calling `rename<L2>`

**Comment by tautschnig** [comment]
> In that case I get the error: `use of deleted function ‘ssa_exprt goto_symex_statet::rename(ssa_exprt, const namespacet&) [with goto_symex_statet::levelt level = (goto_symex_statet::levelt)2u]’`
when calling `rename<L2>`

Hmm, too bad. Maybe we should just not use templates here at all? Just have functions `rename_l0`, `rename_l1`, `rename_l2`? For `rename_l0` and `rename_l1` we would then have polymorphic variants  `exprt -> exprt` and `ssa_exprt -> ssa_exprt`, while for `rename_l2` we would only have `exprt -> exprt`, or maybe `ssa_exprt -> exprt`?

**Comment by romainbrenguier** [comment]
@tautschnig 
> Hmm, too bad. Maybe we should just not use templates here at all? Just have functions rename_l0, rename_l1, rename_l2? For rename_l0 and rename_l1 we would then have polymorphic variants exprt -> exprt and ssa_exprt -> ssa_exprt, while for rename_l2 we would only have exprt -> exprt, or maybe ssa_exprt -> exprt?

The goal in having a template for the non SSA case was to avoid code duplication, because the version for three levels have a lot in common.
I think the current version of the interface is fine, the user can always use `rename<level>`, and the `rename_level0/1_ssa` is there only if they want to avoid using `to_ssa_expr` on the return value (and avoid some unnecessary condition checks in the called function).
I could replace the two functions by a template `rename_ssa<levelt>` if you think that's better (the L2 version would be deleted).

**Comment by tautschnig** [comment]
> I could replace the two functions by a template `rename_ssa<levelt>` if you think that's better (the L2 version would be deleted).

Ok, that's probably the best approach. It makes functions that do the same (though slightly more performant for `ssa_exprt`) (almost) look the same.

**Comment by romainbrenguier** [comment]
@tautschnig @smowton I added a commit to merge the SSA function into one template

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: a3fe46733bb36fb5286020fd328af1bf41c9d392).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/102488874

File: src/goto-symex/goto_symex_state.cpp [hunk 10]
```diff
@@ -330,13 +330,9 @@ void goto_symex_statet::rename(exprt &expr, const namespacet &ns)
...
     if(as_const(expr).type().id() == ID_code)
-    {
       rename<level>(expr.type(), to_symbol_expr(expr).get_identifier(), ns);
-      return;
-    }
-
-    expr=ssa_exprt(expr);
-    rename<level>(expr, ns);
+    else
+      expr = rename<level>(ssa_exprt(expr), ns);
```
Comment (tautschnig) [inline src/goto-symex/goto_symex_state.cpp]: Nit pick: you otherwise aim to use `{...}` in constructor calls, so do `ssa_exprt{expr}` here as well.

File: src/goto-symex/goto_symex_state.cpp [hunk 11]
```diff
@@ -263,24 +263,29 @@ void goto_symex_statet::set_l2_indices(
...
-}
+  static_assert(
+    level == L0 || level == L1,
+    "rename_ssa can only be used for levels L0 and L1");
+  if(level == L0)
+    set_l0_indices(ssa, ns);
+  else if(level == L1)
+    set_l1_indices(ssa, ns);
+  else
+    UNREACHABLE;
```
Comment (tautschnig) [inline src/goto-symex/goto_symex_state.cpp:277]: How about doing a full specialization here? It would do away with the `static_assert` and all the branching (that really isn't one, because the compiler has to remove it anyway). But that's a nit-pick, up to you.

File: src/goto-symex/goto_symex_state.cpp [hunk 12]
```diff
@@ -331,8 +336,10 @@ exprt goto_symex_statet::rename(exprt expr, const namespacet &ns)
     // we never rename function symbols
     if(as_const(expr).type().id() == ID_code)
       rename<level>(expr.type(), to_symbol_expr(expr).get_identifier(), ns);
-    else
+    else if(level == L0 || level == L1)
       expr = rename<level>(ssa_exprt(expr), ns);
+    else // No SSA version for level2
+      expr = rename<level>((exprt)ssa_exprt(expr), ns);
```
Comment (tautschnig) [inline src/goto-symex/goto_symex_state.cpp]: Is this still necessary, given they ssa vs. non-ssa functions have different names? It if is, use `static_cast`, not `(exprt)`.

File: src/goto-symex/symex_function_call.cpp [hunk 13]
```diff
@@ -395,8 +395,8 @@ static void locality(
       it++)
   {
     // get L0 name
-    ssa_exprt ssa =
-      state.rename_level0_ssa(ssa_exprt(ns.lookup(*it).symbol_expr()), ns);
+    ssa_exprt ssa = state.rename_ssa<goto_symex_statet::L0>(
+      ssa_exprt(ns.lookup(*it).symbol_expr()), ns);
```
Comment (tautschnig) [inline src/goto-symex/symex_function_call.cpp]: Use `ssa_exprt{...}`

**Review by tautschnig — APPROVED** [review]

**Comment by tautschnig** [comment]
Any idea about performance differences (if any)?

File: src/goto-symex/goto_symex_state.cpp (hunk 11 above)
Comment (romainbrenguier) [reply src/goto-symex/goto_symex_state.cpp:277]: This version avoids a bit a code duplication, so I would leave it like that. I think the static assert gives an error message that is clearer than getting a linker error.

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 7da918f2b8fd8aefe2acc583828d718da9af35ae).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/102504151

**Comment by romainbrenguier** [comment]
@tautschnig 
> Any idea about performance differences (if any)?

I ran symex on 70 functions, the time before the changes was `102.72` and after `100.88`.

**Review by chrisr-diffblue — APPROVED** [review]
